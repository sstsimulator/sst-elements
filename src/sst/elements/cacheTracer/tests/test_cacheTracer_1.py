# Simple SST python tests to demonstrate use of trace 
## arch model
#
#  comp_cpu <-> comp_tracer <-> comp_l1cache <-> comp_l2cache <-> comp_memory
#
## 

import sst

# Define SST core options
sst.setProgramOption("stop-at", "1ms")

#define simulation components
comp_cpu = sst.Component("cpu0", "memHierarchy.standardCPU")
comp_cpu.addParams({
    "memFreq" : 5,
    "memSize" : "100KiB",
    "verbose" : 0,
    "clock" : "2GHz",
    "rngseed" : 111,
    "maxOutstanding" : 16,
    "opCount" : 1000,
    "reqsPerIssue" : 2,
    "write_freq" : 35, # 35% writes
    "read_freq" : 65,  # 65% reads
})

iface = comp_cpu.setSubComponent("memory", "memHierarchy.standardInterface")

comp_l1cache = sst.Component("l1cache", "memHierarchy.Cache")
comp_l1cache.addParams({
    "access_latency_cycles" : "5",
    "cache_frequency"       : "2 Ghz",
    "replacement_policy"    : "lru",
    "coherence_protocol"    : "MSI",
    "associativity"         : "4",
    "cache_line_size"       : "64",
    "debug_level"           : "8",
    "L1"                    : "1",
    "debug"                 : "0",
    "cache_size"            : "4 KB",
})

comp_l2cache = sst.Component("l2cache", "memHierarchy.Cache")
comp_l2cache.addParams({
    "access_latency_cycles" : "20",
    "cache_frequency"       : "2 Ghz",
    "replacement_policy"    : "lru",
    "coherence_protocol"    : "MSI",
    "associativity"         : "4",
    "cache_line_size"       : "64",
    "debug_level"           : "8",
    "L1"                    : "0",
    "debug"                 : "0",
    "cache_size"            : "64 KB",
})

comp_memory = sst.Component("memory", "memHierarchy.MemController")
comp_memory.addParams({
    "clock"                 : "2 Ghz",
    "request_width"         : "64",
    "debug"                 : "0",
})
    
backend = comp_memory.setSubComponent("backend", "memHierarchy.simpleMem")
backend.addParams({ "mem_size"      : "1024MiB" })

comp_tracer = sst.Component("tracer", "cacheTracer.cacheTracer")
comp_tracer.addParams({
    "clock"      : "2 Ghz", 
    "debug"      : "0",
    "statistics" : "1",
    "tracePrefix" : "",
    "statsPrefix" : "",
    "pageSize"   : "4096",
    "accessLatencyBins" : "10",
 })

# define the simulation links
link_cpu_tracer = sst.Link("link_cpu_tracer")
link_cpu_tracer.connect((iface, "lowlink", "100ps"), (comp_tracer, "northBus", "100ps"))

link_tracer_l1cache = sst.Link("link_tracer_l1cache")
link_tracer_l1cache.connect((comp_tracer, "southBus", "100ps"), (comp_l1cache, "highlink", "100ps"))

link_l1cache_l2cache = sst.Link("link_l1cache_l2cache")
link_l1cache_l2cache.connect((comp_l1cache, "lowlink", "100ps"), (comp_l2cache, "highlink", "100ps"))

link_l2cache_mem = sst.Link("link_l2cache_mem")
link_l2cache_mem.connect((comp_l2cache, "lowlink", "100ps"), (comp_memory, "highlink", "100ps"))

