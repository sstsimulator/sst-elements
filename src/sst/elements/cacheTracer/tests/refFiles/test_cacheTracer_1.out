l2cache: No MSHR lookup latency provided (mshr_latency_cycles)...intrapolated to 5 cycles.
memory, WARNING: Memories no longer inherit address regions from directories and no region parameters (addr_range_start, addr_range_end, interleave_size, interleave_step) were detected. All addresses will map to this memory: if this is intended, you may ignore this warning or set addr_range_start to 0 in your input deck to eliminate this warning.
TrivialCPU: Test Completed Successfuly
FINAL STATS:
-----------------------------------------------------------------
- Events at NorthBus                 : 1000
- Events at SouthBus                 : 1000
- Total Events                       : 2000
-----------------------------------------------------------------

Address Histogram:
-----------------------------------------------------------------
Address_Range: Count
- [0-4095]: 70
- [4096-8191]: 62
- [8192-12287]: 66
- [12288-16383]: 60
- [16384-20479]: 64
- [20480-24575]: 60
- [24576-28671]: 58
- [28672-32767]: 73
- [32768-36863]: 66
- [36864-40959]: 57
- [40960-45055]: 74
- [45056-49151]: 60
- [49152-53247]: 59
- [53248-57343]: 59
- [57344-61439]: 55
- [61440-65535]: 57
-----------------------------------------------------------------
- Total_Events_Address: 1000
-----------------------------------------------------------------

Access Latency Distribution (ns):
-----------------------------------------------------------------
Min-Latency(ns): 4  Max-Latency(ns): 131  #Bins: 10
-----------------------------------------------------------------
Latency Range(ns): Count
- [0-13]: 63
- [14-27]: 320
- [28-41]: 0
- [42-55]: 0
- [56-69]: 0
- [70-83]: 0
- [84-97]: 0
- [98-111]: 0
- [112-125]: 616
- [126-139]: 1
-----------------------------------------------------------------
- Total_Events_Latency: 1000
-----------------------------------------------------------------

TrivialCPU cpu0 Finished after 1000 issued reads, 1000 returned (104054 clocks)
Simulation is complete, simulated time: 104.054 us
